### ISSUE TITLE
Join forces with Spacemacs

### ISSUE THREAD

**Issue by Ambrevar** [emacs-evil/evil-collection#24]
Spacemacs community is _huge_ (apparently much bigger than I initially thought) and diverging too much from it could be a terrible move resulting in more fragmentation that we would ever dread.
On the upside, we would clearly benefit a lot from such a knowledge base and a community of this size.

This awareness was brought to me in the following discussion:
https://www.reddit.com/r/emacs/comments/7brlns/need_opinions_could_it_be_worthwhile_to_switch_to/dpld660/?utm_name=emacs.

As a result I've contacted Spacemacs maintainers to see what we can do about that:
https://github.com/syl20bnr/spacemacs/issues/9850.

I've update the README with a link to Spacemacs' conventions (i.e. rationale):
https://github.com/syl20bnr/spacemacs/blob/master/doc/CONVENTIONS.org#key-bindings-conventions.

**Comment by bluewww** [comment]
Spacemacs is one leading factors for emacs growth thats why, in my opinion, we absolutely must be compatible (shortcut wise) with it. The spacemacs shortcut convention is definitely a good start.

**Comment by Ambrevar** [comment]
What do you mean precisely with "shortcut convention"?

While there is definite room for collaboration, I don't think evil-collection should adhere to the `SPC m` approach of Spacemacs.

From the [Spacemacs convention](https://github.com/syl20bnr/spacemacs/blob/master/doc/CONVENTIONS.org#code-navigation)
```
SPC m g a	go to alternate file (i.e. .h <--> .cpp)
SPC m g b	go back to previous location (before last jump)
SPC m g g	go to things under point
SPC m g G	go to things under point in other window
SPC m g t	go to corresponding test file if any
```

In my opinion, those bindings are too long for something that is used so often.
The `g` prefix can definitely be exploited here, like the popular `gd` for "[g]o to [d]efinition".

I think all `SPC m` bindings can be shortened by one or two keys at least.

**Comment by Ambrevar** [comment]
> Spacemacs is one leading factors for emacs growth thats why, in my opinion, we absolutely must be compatible (shortcut wise) with it.

It's a good point, but on the other hand, Emacs development is not (and should not) be directed by Spacemacs specificities.

**Comment by bluewww** [comment]
SPC m (major mode leader) is the same as ",", so I dont think those key bindings are too long.

**Comment by Ambrevar** [comment]
> it might just cause chaos down the road.

Which is why the centrepiece of this project is the rationale: ensure consistency across all modes.

In practice this is working out quite well so far with the dozens of modes we've
already covered.

> I also think that you shouldnt bind "SPC" to things since that is a nice modifier key.

Any key can be a modifier key, we cannot cripple the bindings to satisfy everyone and the 17 modifier / leader keys used out there.
See #5 for a possibly better solution.

> Shortcut is a different word for key binding.

I just wanted to make sure you meant the same CONVENTION.org and not something else.

**Comment by jojojames** [comment]
There seems to be a few opposing forces at play here.

1. The package is primarily intended at users that want to use evil but don't want to use Spacemacs.
It feels a little out of scope to support it.

@bluewww, What value do you see in this project if you're already using Spacemacs?

2. Spacemacs has a different keybinding philosophy, they heavily rely on two things.

SPC has a leader key with a long set of pneumonic keybindings is where most keys go and evilify-map, a generic way to get evil keys in modes that don't have evil keys in them.

We don't support the former and the latter is at odds with how we want to bind keys (evilify vs evil-define-key).

Because of the emphasis on the leader key, I think the resulting evil keybindings suffer (since they are generically created) or aren't as good as they could be.

3. I'm not sure chaining this project to another is the right idea if we want velocity. It's only been a few days, but @Ambrevar and I are making good progress. I think very soon (if not already/I don't use Spacemacs), we'll have more varied support for modes than Spacemacs does.

I think there's value in fragmentation if that fragmentation leads to supporting different ideaologies.

@Ambrevar You seem to be more open to the idea so I'll continue thinking about the topic but my thinking currently is that it's probably not a good idea.

**Comment by Ambrevar** [comment]
@jojojames: I actually agree with all your points, but consider this: https://github.com/syl20bnr/spacemacs/issues/9850#issuecomment-343526993.  As you said, there is an inherent philosophical issue with Spacemacs vs. upstream constribution.
I'm afraid the idea of joining our efforts won't be fruitful unless Spacemacs officially steers towards a more upstream-oriented philosophy.
In that case, their might be some valuable contribution happening.

Regardless of all this, the `SPC m` leader approach is a huge design difference with evil-collection: I do not know how we can connect the dots here.
If nobody can think of a solution to accommodate the universal bindings over the `SPC m` design, then I'm afraid this will be a dead-end.

**Comment by noctuid** [comment]
> Spacemacs is one leading factors for emacs growth thats why, in my opinion, we absolutely must be compatible (shortcut wise) with it. 

I agree with @jojojames. I don't really see where the room for compatibility would be. `evil-collection` doesn't address global keybindings, and spacemacs' `evilify` doesn't make any attempt to give consistent keybindings across modes. The `SPC m` keybindings generally only make sense in normal/editable modes.

**Comment by Ambrevar** [comment]
Let's wait and see where the discussion goes on Spacemacs' side.
In the meantime, let's keep up with the good work.

**Comment by Ambrevar** [comment]
Considering the [take of Spacemacs' lead on the issue](https://github.com/syl20bnr/spacemacs/issues/9850#issuecomment-343749828), a collaboration is unlikely to happen.  I suggest we wait a few weeks after the release on MELPA before closing this issue.

**Comment by Ambrevar** [comment]
I guess this won't move forward so I'm closing this for now.

**Comment by jojojames** [comment]
@Ambrevar I was browsing through the old issue on spacemacs and almost missed a post from @syl20bnr.

Here is the quoted post: (Putting it here for visibility)

```
Recent events made me think about it a bit more, I drop this here if someone wants to tackle this.

Spacemacs could use an external package for general evil key bindings integration but this package must have the following design:

    being non destructive: map eveything in a special evil state
    apply clearly and easy to understand rules to auto-remap: auto-evilification
    fine-tune the rest of the key bindings if needed
    loading system that plays nicely with isolated configuration design like the layer system
    easy customization for fine-tuning

evil-collection surely checks some of the points above but since it is not powered under the hood by an automatic evilification system with isolated evil state I'm not really interested in it as I fear it can lead to conventions tedious to maintain and hard to enforce correctly.

See (complicated evasive guideline, i.e. when possible) https://github.com/emacs-evil/evil-collection#guidelines vs. (always applied conventions with consistent rules) https://github.com/syl20bnr/spacemacs/blob/develop/doc/CONVENTIONS.org#evilified-buffers
```

And replying individually with my thoughts:

> being non destructive: map eveything in a special evil state

Hmnn... Don't know what to think here, initial thoughts leads me to just preferring normal state.. But I could be convinced.

>    apply clearly and easy to understand rules to auto-remap: auto-evilification

I'm not a fan of autoremapping keys. I think each mode has its own quirks and gotchas that make it more effective to handpick the right keys for each. Given enough community contribution, at some point there will be a more solid/handcrafted approach to the mappings. Autoremapping (from what I've seen in Spacemacs) is pretty slick but leads me to typing keys that I wouldn't have wanted to type given any specific mode. (Imagine dired remapped)

 >   fine-tune the rest of the key bindings if needed

Agree.

>    loading system that plays nicely with isolated configuration design like the layer system

I think we'd need to know how the current system would be incompatible with the layering system Spacemacs has first before we can come up with a way to play nice.

>    easy customization for fine-tuning

Agree.

> evil-collection surely checks some of the points above but since it is not powered under the hood by an automatic evilification system with isolated evil state I'm not really interested in it as I fear it can lead to conventions tedious to maintain and hard to enforce correctly.

I think mostly everyone just wants 80/20, (i.e. for most of their keys to be mapped reasonably). We can continue polishing each mode's keymaps as it goes and with more contributors, I think eventually handrolled remappings will have a more polished experience than an automated version.

The isolated evil state is another big one. I personally am not a fan of 'motion so seeing another evil state gives me pause.

**Comment by jojojames** [comment]
Leaving as closed for now but feel free to reopen anyone.

**Comment by eeshugerman** [comment]
new developments: https://github.com/syl20bnr/spacemacs/pull/14200

**Comment by jojojames** [comment]
Thanks for the notice @eeshugerman , added a comment in that thread re additional bindings they may need to make.

Re; further integration, etc: I don't think much has changed other than spacemacs begrudgingly pulling in evil-collection for magit.

The original points from @syl20bnr still hold and they are at odds with evil-collection.

Reposted here for clarity:

```
    being non destructive: map eveything in a special evil state
    apply clearly and easy to understand rules to auto-remap: auto-evilification
    fine-tune the rest of the key bindings if needed
    loading system that plays nicely with isolated configuration design like the layer system
    easy customization for fine-tuning
```

Would be happy to join forces but with the two projects directions going in the opposite directions, it's probably not viable (manual key definitions vs auto evilification being the big one).
